src/base/packing_stream.cpp

```diff
@@ -0,0 +1,564 @@
+/*
+ * Copyright (C) 2008 Search Solution Corporation. All rights reserved by Search Solution.
+ *
+ *   This program is free software; you can redistribute it and/or modify
+ *   it under the terms of the GNU General Public License as published by
+ *   the Free Software Foundation; either version 2 of the License, or
+ *   (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program; if not, write to the Free Software
+ *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+ *
+ */
+
+/*
+ * packing_stream.cpp
+ */
+
+#ident "$Id$"
+
+#include "packing_stream.hpp"
+
+#include "error_code.h"
+#include "error_manager.h"
+
+namespace cubstream
+{
+  packing_stream::packing_stream (const size_t buffer_capacity, const int max_appenders)
+    : m_bip_buffer (buffer_capacity),
+      m_reserved_positions (max_appenders)
+  {
+    m_oldest_readable_position = 0;
+
+    /* TODO : system parameter */
+    m_trigger_flush_to_disk_size = buffer_capacity / 2;
+
+    m_trigger_min_to_read_size = 1024;
+
+    m_stat_reserve_queue_spins = 0;
+    m_stat_reserve_buffer_spins = 0;
+    m_stat_read_not_enough_data_cnt = 0;
+    m_stat_read_not_in_buffer_cnt = 0;
+    m_stat_read_no_readable_pos_cnt = 0;
+    m_stat_read_buffer_failed_cnt = 0;
+  }
+
+  packing_stream::~packing_stream ()
+  {
+    assert (m_append_position - m_read_position == 0);
+  }
+
+  /*
+   * write API for stream:
+   * 1. reserve position (with pointer in bip_buffer)
+   * 2. call write action
+   * 3. commit the reserve position
+   *  the write_function is expected to return error code (negative value) or number of written bytes (positive)
+   */
+  int packing_stream::write (const size_t byte_count, write_func_t &write_action)
+  {
+    int err = NO_ERROR;
+    stream_reserve_context *reserve_context = NULL;
+    char *ptr;
+    stream_position reserved_pos;
+    int written_bytes;
+
+    ptr = reserve_with_buffer (byte_count, reserve_context);
+    if (ptr == NULL)
+      {
+	err = ER_FAILED;
+	return err;
+      }
+    reserved_pos = reserve_context->start_pos;
+
+    written_bytes = write_action (reserved_pos, ptr, byte_count);
+    if (written_bytes < 0)
+      {
+	return ER_FAILED;
```

Comment:
I don't think it is safe to return error. It will leave stream with uncommitted write.

What do you think of removing return error from read/write completely. Stream cannot manage errors inside actions these actions. If errors are possible, their management should be entirely caller's responsibility.